tools/format.ts

```diff
@@ -0,0 +1,249 @@
+      {
+        args: [executable, "format", filename]
+      }
+    ]);
+  }
+
+  return processOptions;
+}
+
+function formatPython(): ProcessOptions {
+  const executable = "python";
+  const yapfPath = pathJoin("third_party", "python_packages", "bin", "yapf");
+
+  const pythonFiles = findFiles({
+    dir: ["tools", "build_extra"],
+    ext: ".py",
+    skip: ["tools/clang"]
+  });
+  // TODO: python_env() equivalent (eg in third_party.py)
+  const options: RunOptions = {
+    args: [executable, yapfPath, "-i", ...pythonFiles]
+  };
+
+  return ["yapf", options];
+}
+
+function formatPrettier(): ProcessOptions {
+  const executable = pathJoin(
+    cwd(),
+    "third_party",
+    "node_modules",
+    ".bin",
+    "prettier"
+  );
+
+  const prettierFiles = [
+    ...findFiles({
+      dir: cwd(),
+      ext: [".json", ".md", ".js"],
+      depth: 0
+    }),
+    ...findFiles({
+      dir: [".github", "js", "tests", "tools", "website"],
+      ext: [".js", ".json", ".ts", ".md"],
+      skip: ["tools/clang", "js/deps"]
+    })
+  ];
+  const flags = ["--write" /* "--loglevel=error" */];
+  const options: RunOptions = {
+    args: [executable, ...flags, ...prettierFiles]
+  };
+
+  return ["prettier", options];
+}
+
+function formatRust(): ProcessOptions {
+  const toolsPath = pathJoin(cwd(), "tools");
+  const executable = pathJoin("third_party", "rustfmt", platform.os, "rustfmt");
+  const flags = ["--config-path", pathJoin(toolsPath, "rustfmt.toml")];
+  const rustFilepaths = ["build.rs", ...findFiles({ dir: "src", ext: ".rs" })];
+  const options: RunOptions = {
+    args: [executable, ...flags, ...rustFilepaths]
+  };
+  return ["rustfmt", options];
+}
+
+async function resolveProcess([name, runOpts]: ProcessOptions) {
+  const process: Process = runp(runOpts);
+  const status = await process.status();
+  const failed = status.code !== 0 || status.success !== true;
+
+  if (failed) {
+    const stdout = await readAll(process.stdout);
+    const stderr = await readAll(process.stderr);
+    const message = [
+      `✖ ${name}`,
+      new TextDecoder().decode(stdout).trim(),
+      new TextDecoder().decode(stderr).trim()
+    ]
+      .filter(Boolean)
+      .join("\n");
+
+    return {
+      success: false,
+      message
+    };
+  }
+  return {
+    success: true,
+    message: `✔ ${name}`
+  };
+}
+
+async function main() {
+  const processOptions: ProcessOptions[] = [
+    formatClang(),
+    ...formatGN(),
+    formatPython(),
+    formatPrettier(),
+    formatRust()
+  ];
+
+  const results = await Promise.all(processOptions.map(resolveProcess));
+
+  // Print error messages last
+  const sortedResults = results.sort(
+    (a, b) => Number(b.success) - Number(a.success)
+  );
+
+  for (const { message } of sortedResults) {
+    console.log(message);
+  }
+
+  if (sortedResults[sortedResults.length - 1].success !== true) {
+    return Promise.reject();
+  }
+}
+
+main().catch(err => {
+  if (err) {
+    console.log(err);
+  }
```

Comment:
The catch was mainly to accomodate for not logging out `undefined` to the console for the empty promise rejection. It looks like the way this is handled in other deno scripts is to just call `exit()` rather than returning an empty Promise rejection in main so i’ve done that. EG https://github.com/denoland/deno_std/blob/master/examples/gist.ts#L25